src/scvi/external/sysvi/_base_components.py

```diff
@@ -0,0 +1,396 @@
+            if n_cat and cat is None:
+                raise ValueError("cat not provided while n_cat != 0 in init. params.")
+            if n_cat > 1:  # n_cat = 1 will be ignored - no additional info
+                if cat.size(1) != n_cat:
+                    one_hot_cat = nn.functional.one_hot(cat.squeeze(-1), n_cat)
+                else:
+                    one_hot_cat = cat  # cat has already been one_hot encoded
+                one_hot_cat_list += [one_hot_cat]
+        for i, layers in enumerate(self.fc_layers):
+            for layer in layers:
+                if layer is not None:
+                    if isinstance(layer, nn.BatchNorm1d):
+                        if x.dim() == 3:
+                            x = torch.cat([(layer(slice_x)).unsqueeze(0) for slice_x in x], dim=0)
+                        else:
+                            x = layer(x)
+                    else:
+                        if isinstance(layer, nn.Linear) and self.inject_into_layer(i):
+                            if x.dim() == 3:
+                                cov_list_layer = [
+                                    o.unsqueeze(0).expand((x.size(0), o.size(0), o.size(1)))
+                                    for o in one_hot_cat_list
+                                ]
+                            else:
+                                cov_list_layer = one_hot_cat_list
+                            x = torch.cat((x, *cov_list_layer, *cont_list), dim=-1)
+                        x = layer(x)
+        return x
+
+
+class VarEncoder(Module):
+    """Encode variance (strictly positive).
+
+    Parameters
+    ----------
+    n_input
+        Number of input dimensions.
+        Used if mode is ``'sample_feature'`` to construct a network predicting
+        variance from input features.
+    n_output
+        Number of variances to predict, matching the desired number of features
+        (e.g. latent dimensions for variational encoding or output features
+        for variational decoding).
+    mode
+        How to compute variance.
+        One of the following:
+        * ``'sample_feature'`` - learn variance per sample and feature.
+        * ``'feature'`` - learn variance per feature, constant across samples.
+    """
+
+    def __init__(
+        self,
+        n_input: int,
+        n_output: int,
+        mode: Literal["sample_feature", "feature", "linear"],
+        activation: Callable | None = None,
+    ):
+        super().__init__()
+
+        self.clip_exp = np.log(torch.finfo(torch.get_default_dtype()).max) - 1e-4
+        self.mode = mode
+        if self.mode == "sample_feature":
+            self.encoder = Linear(n_input, n_output)
+        elif self.mode == "feature":
+            self.var_param = Parameter(torch.zeros(1, n_output))
+        else:
+            raise ValueError("Mode not recognised.")
+        self.activation = torch.exp if activation is None else activation
```

Comment:
Do you still need nan_to_num if you use here a softplus activation? You should consider changing it. It's just for backward compatibility still exp in scvi.